### PR TITLE
NameResolver can now produce addresses without ports

### LINKED ISSUE #1138: Propagate absent port from NameResolver to ManagedChannelImpl
When a target URI doesn't contain a port, NameResolver should propagate this absence to ManagedChannelImpl, who can based on encryption is in use or not, select 443 or 80 as the actual port.

### PR DESCRIPTION
**Pull request by zhangkun83**
- InetSocketAddress with port 0 means port not specified, in which case
  ManagedChannelImpl will use the default port passed from channel
  builders.
- Channel builders decide the default port based on whether TLS is used.
- Channel builders for which port is not applicable, e.g., InProcessChannelBuilder, will return -1 as the default port.

Resolves #1138

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java [hunk 1]
```diff
@@ -175,9 +175,17 @@ public ManagedChannelImpl build() {
    * {@link ClientTransportFactory} appropriate for this channel.  This method is meant for
    * Transport implementors and should not be used by normal users.
    */
-  @Internal
   protected abstract ClientTransportFactory buildTransportFactory();
 
+  /**
+   * Children of AbstractChannelBuilder should override this method to provide the default port for
+   * creating outgoing connections in case the {@link NameResolver} not provide a port.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: s/not/does not/

File: core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java [hunk 2]
```diff
@@ -175,9 +175,17 @@ public ManagedChannelImpl build() {
    * {@link ClientTransportFactory} appropriate for this channel.  This method is meant for
    * Transport implementors and should not be used by normal users.
    */
-  @Internal
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: Did you intend to remove this internal annotation. Maybe you want to make it `@ExperimantalApi`?

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: Oh. nm. I see. This is an internal class, so it is redundant.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java]: This class is under internal package, so the annotation is superfluous.

**Comment by ejona86** [comment]
@zhangkun83 LGTM.

I sort of question if maybe it would be better like adding a parameter to `NameResolverFactory.newNameResolver()` for providing a default port. It could even be Attributes so that it could be extended easily.

But this seems fine.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 3]
```diff
@@ -185,7 +189,21 @@
...
+          SocketAddress address = server.getAddress();
+          if (address instanceof InetSocketAddress) {
+            InetSocketAddress inetSocketAddress = (InetSocketAddress) address;
+            if (inetSocketAddress.getPort() == 0) {
+              server = new ResolvedServerInfo(
+                  new InetSocketAddress(
+                      inetSocketAddress.getAddress(), ManagedChannelImpl.this.defaultPort),
+                  server.getAttributes());
+            }
+          }
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: What should happen in address is not an InetSocketAddr?

**Comment by zhangkun83** [comment]
> I sort of question if maybe it would be better like adding a parameter to NameResolverFactory.newNameResolver() for providing a default port. It could even be Attributes so that it could be extended easily.

@ejona86 I think it's a good suggestion. It would avoid the special accommodation for InetSocketAddress, and the special port 0 in InetSocketAddress. I have created #1158. If that one looks good to you, I will abandon this one.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 3 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: This is not ideal. Based on @ejona86's suggestion, I have created #1158 and will abandon this one.

**Pull request closed without merging**